lib/ansible/modules/source_control/git.py

```diff
@@ -977,6 +1004,59 @@ def main():
     track_submodules = module.params['track_submodules']
 
     result.update(before=None)
+
+    if archive:
+        if module.check_mode:
+            result.update(changed=True)
+            module.exit_json(**result)
+
+        all_archive_fmt = {'.zip': 'zip', '.gz': 'tar.gz', '.tar': 'tar',
+                           '.tgz': 'tgz'}
+        _, archive_ext = os.path.splitext(archive)
+        archive_fmt = all_archive_fmt.get(archive_ext, None)
+        if archive_fmt is None:
+            module.fail_json(msg="Unable to get file extension from "
+                                 "archive file name : %s" % archive,
+                             details="Please specify archive as filename with "
+                                     "extension. File extension can be one "
+                                     "of ['tar', 'tar.gz', 'zip', 'tgz']")
+
+        repo_name = repo.split("/")[-1].replace(".git", "")
+        # Git archive is not supported by all git servers, so
+        # we will first clone and perform git archive from local directory
+        if dest and not os.path.exists(gitconfig):
+            clone(git_path, module, repo, dest, remote, depth, version, bare,
```

Comment:
I meant at the end of the module, so after https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/source_control/git.py#L1067

At that point you can assume dest exists and has the version you want checked out, so you can remove this if/clone statement and just move the rest of the code down.

Since the main routine is quite long it's probably a good idea to put all the archive code in a separate routine, such that in main you have something like
```
if archive:
    create_archive(module, archive, git_path, version)
```